aws-xray/src/test/java/io/opentelemetry/contrib/awsxray/AwsAttributePropagatingSpanProcessorTest.java

```diff
@@ -0,0 +1,147 @@
+    validateSpanAttributesInheritance(spanWithAppAndOp, null, "testService", "testOperation");
+  }
+
+  @Test
+  public void testOverrideRemoteAttributes() {
+    Span parentSpan = tracer.spanBuilder("parent").startSpan();
+    parentSpan.setAttribute(AwsAttributeKeys.AWS_REMOTE_SERVICE, "testService");
+    parentSpan.setAttribute(AwsAttributeKeys.AWS_REMOTE_OPERATION, "testOperation");
+
+    Span transmitSpans1 = createNestedSpan(parentSpan, 2);
+
+    Span childSpan =
+        tracer.spanBuilder("child:1").setParent(Context.current().with(transmitSpans1)).startSpan();
+
+    childSpan.setAttribute(AwsAttributeKeys.AWS_REMOTE_OPERATION, "childOperation");
+
+    Span transmitSpans2 = createNestedSpan(childSpan, 2);
+
+    assertThat(((ReadableSpan) transmitSpans2).getAttribute(AwsAttributeKeys.AWS_REMOTE_OPERATION))
+        .isEqualTo("childOperation");
+  }
+
+  @Test
+  public void testRemoteAttributesNotExists() {
+    Span span = tracer.spanBuilder("parent").startSpan();
+    validateSpanAttributesInheritance(span, null, null, null);
+  }
+
+  @Test
+  public void testLocalAttributeDetectionBySpanKind() {
+    for (SpanKind value : SpanKind.values()) {
+      Span span = tracer.spanBuilder("startOperation").setSpanKind(value).startSpan();
+      if (value == SpanKind.SERVER || value == SpanKind.CONSUMER) {
+        validateSpanAttributesInheritance(span, "startOperation", null, null);
+      } else {
+        validateSpanAttributesInheritance(span, null, null, null);
+      }
+    }
+  }
+
+  @Test
+  public void testLocalAttributeDetectionWithRemoteParentSpan() {
+    Span remoteParent =
+        Span.wrap(
+            SpanContext.createFromRemoteParent(
+                "00000000000000000000000000000001",
+                "0000000000000002",
+                TraceFlags.getSampled(),
+                TraceState.getDefault()));
+    Context parentcontext = Context.root().with(remoteParent);
+    Span span =
+        tracer
+            .spanBuilder("startOperation")
+            .setSpanKind(SpanKind.SERVER)
+            .setParent(parentcontext)
+            .startSpan();
+    validateSpanAttributesInheritance(span, "startOperation", null, null);
+  }
+
+  private Span createNestedSpan(Span parentSpan, int depth) {
+    if (depth == 0) {
+      return parentSpan;
+    }
+    Span childSpan =
+        tracer
+            .spanBuilder("child:" + depth)
+            .setParent(Context.current().with(parentSpan))
+            .startSpan();
+    try {
+      return createNestedSpan(childSpan, depth - 1);
+    } finally {
+      childSpan.end();
+    }
+  }
+
+  private void validateSpanAttributesInheritance(
+      Span spanWithAppOnly, String localOperation, String remoteService, String remoteOperation) {
+    ReadableSpan leafSpan = (ReadableSpan) createNestedSpan(spanWithAppOnly, 10);
+
+    assertThat(leafSpan.getParentSpanContext()).isNotNull();
+    assertThat(leafSpan.getName()).isEqualTo("child:1");
```

Comment:
If the `leafSpan` is the first child of the `spanWithAppOnly`, then what's the point of creating 10 nested spans in the `createNestedSpan`. I think the purpose is to test the propagation to the deepest span, right?

Also, `spanWithAppOnly` should be renamed to `parentSpan` as this is a generic utility method.

Reply:
Looking at `createNestedSpan`, the returned span is the deepest span. The number in the span name decreases at each level so child `depth` is the first child and child 1 is leaf.

Yea, can rename variable.

Reply:
Ah, I see. Thanks for clarifying. :)